activation/e2e/nipost_test.go

```diff
@@ -176,173 +187,44 @@ func TestNIPostBuilderWithClients(t *testing.T) {
 
 	t.Cleanup(launchPostSupervisor(t, logger, mgr, sig, grpcCfg, opts))
 
+	var postClient activation.PostClient
 	require.Eventually(t, func() bool {
-		_, err := svc.Client(sig.NodeID())
+		var err error
+		postClient, err = svc.Client(sig.NodeID())
 		return err == nil
 	}, 10*time.Second, 100*time.Millisecond, "timed out waiting for connection")
 
-	localDB := localsql.InMemory()
-	nb, err := activation.NewNIPostBuilder(
-		localDB,
-		poetDb,
-		svc,
-		[]types.PoetServer{{Address: poetProver.RestURL().String()}},
-		logger.Named("nipostBuilder"),
-		poetCfg,
-		mclock,
-	)
-	require.NoError(t, err)
-
-	challenge := types.NIPostChallenge{
-		PublishEpoch: postGenesisEpoch + 2,
-	}
-	nipost, err := nb.BuildNIPost(context.Background(), sig, &challenge)
+	post, info, err := postClient.Proof(context.Background(), shared.ZeroChallenge)
 	require.NoError(t, err)
 
-	v := activation.NewValidator(nil, poetDb, cfg, opts.Scrypt, verifier)
-	_, err = v.NIPost(
-		context.Background(),
-		sig.NodeID(),
-		goldenATX,
-		nipost.NIPost,
-		challenge.Hash(),
-		nipost.NumUnits,
-	)
-	require.NoError(t, err)
-}
-
-func TestNIPostBuilder_Close(t *testing.T) {
-	ctrl := gomock.NewController(t)
-
-	sig, err := signing.NewEdSigner()
-	require.NoError(t, err)
-
-	logger := zaptest.NewLogger(t)
-
-	poetProver := spawnPoet(t, WithGenesis(time.Now()), WithEpochDuration(time.Second))
-	poetDb := activation.NewMockpoetDbAPI(ctrl)
-
-	mclock := activation.NewMocklayerClock(ctrl)
-	mclock.EXPECT().LayerToTime(gomock.Any()).AnyTimes().DoAndReturn(
-		func(got types.LayerID) time.Time {
-			// time.Now() ~= currentLayer
-			genesis := time.Now().Add(-time.Duration(postGenesisEpoch.FirstLayer()) * layerDuration)
-			return genesis.Add(layerDuration * time.Duration(got))
-		},
-	)
-
-	svc := grpcserver.NewPostService(logger)
-
-	db := localsql.InMemory()
-	nb, err := activation.NewNIPostBuilder(
-		db,
-		poetDb,
-		svc,
-		[]types.PoetServer{{Address: poetProver.RestURL().String()}},
-		logger.Named("nipostBuilder"),
-		activation.PoetConfig{},
-		mclock,
+	client, err := activation.NewHTTPPoetClient(
+		types.PoetServer{Address: poetProver.RestURL().String()},
+		poetCfg,
+		activation.WithLogger(logger),
 	)
 	require.NoError(t, err)
 
-	challenge := types.NIPostChallenge{
-		PublishEpoch: postGenesisEpoch + 2,
-	}
-	ctx, cancel := context.WithCancel(context.Background())
-	cancel()
-	nipost, err := nb.BuildNIPost(ctx, sig, &challenge)
-	require.ErrorIs(t, err, context.Canceled)
-	require.Nil(t, nipost)
-}
-
-func TestNewNIPostBuilderNotInitialized(t *testing.T) {
```

Comment:
This test also had a purpose. I'm not sure we still need it, but it tested that if PoST is not initialised while the `activation.Builder`/`nipostBuilder` were already querying for PoSTs it would not result in generating invalid posts but rather the builder waiting / retrying until PoST has finished initializing.

Reply:
This test didn't do what you described. There is a sequence:

```go
	t.Cleanup(launchPostSupervisor(t, logger, mgr, sig, grpcCfg, opts))

	require.Eventually(t, func() bool {
		_, err := svc.Client(sig.NodeID())
		return err == nil
	}, 10*time.Second, 100*time.Millisecond, "timed out waiting for connection")

	challenge := types.RandomHash()
	nipost, err := nb.BuildNIPost(context.Background(), sig, 7, challenge)
```

The `nb.BuildNIPost()` is called after initialization finished (after the post service already connected, and the service is only started when init finished).

I removed it because I found it unneccessary/duplicated.

Reply:
Even if I remove waiting for the post service to connect, is not guaranteed to test this. Most likely, the initialization will finish before poet and the `nipostBuilder` will try to build PoST when the init is already done.

Reply:
No the waiting is required. The difference between this test and e.g. `TestNIPostBuilderWithClients` is that `TestNIPostBuilderWithClients` calls `initPost` before `launchPostSupervisor`; so the supervisor is started after the post data has already been completely generated.

This test did not do this, it just launched the Post Supervisor to ensure it will generate the data itself before actually starting the post service and waiting for the connection to the grpc service.